### ISSUE TITLE
Contexts/Callbacks/Hook ideas (RFC)

### ISSUE THREAD

**Issue by mieko** [rails/actioncable#32]
I'm porting an application from Rails 4 + a homegrown Faye mess to Rails 5-pre + ActionCable.  I was probably the most excited person on earth when this repo went public.

This RFC is about Connection/Channel lifecycle hooks generally, but I'll illustrate with my actual use-case.

The application is heavily tenanted, with a model that works a lot like the Apartment gem.  Outside of a few well-defined areas, all of our logic looks like normal Rails practices.  Models, Controllers, Jobs, etc, are written as if nothing weird with tenancy is going on behind the scenes.  This is possible because there tends to be a publicly-documented bottleneck I can use to swap the ground beneath the code for each of the frameworks, things like this: https://gist.github.com/mieko/13a312b10b6087f38a87

I'm trying to find a clean way to extend this behavior to ActionCable.  To cover all cases, I've had to end up with something like the following (paraphrased for clarity):

```ruby
module ApplicationCable
  class Connection < ActionCable::Connection::Base
    include ActivateTenant
    identified_by :current_tenant, :current_user

    def connect
      # Pull the tenant out of the hostname
      self.current_tenant = tenant_for_request(request)
      self.current_user = user_for_request(request)
    end
  end
end
```

And here's where it gets hacky:

```ruby
module ApplicationCable
  class Channel < ActionCable::Channel::Base
    # For sake of exposition, each method written out 
    protected
    def dispatch_action(*); current_tenant.activate! { super }; end
    def run_subscribe_callbacks(*); current_tenant.activate! { super }; end
    def run_unsubscribe_callbacks(*); current_tenant.activate! { super }; end
  end
end
```

This seems(?) to be required because ActionCable doesn't have the "full-featured" callback handling of something like ActionController.  `on_{un,}subscribe` would be perfect for this, but AC::C::Base registers `subscribed` itself for instances, and registering further callbacks append to the chain.  This means `ApplicationCable::Channel` cannot register an `on_subscribe` that'd execute before subclass instances' `subscribed`.

Without going down the rabbit hole of `prepend_...` and `append_...`, I think it'd be nice if something like the following were to work out of the box:

```ruby
module ApplicationCable
  class Channel < ActionCable::Channel::Base
    identified_by ...
    on_subscribe :activate_tenant
    
    protected
    def activate_tenant
      current_tenant.activate!(...)
    end
  end
end

class AppearanceChannel < ApplicationCable::Channel
  def subscribed
    # This executes after the on_subscribe registered by the parent class, without
    # being careful to super in each channel.
  end
end
```

Basically running in Base -> Subclass order for `on_subscribe`, and I guess normal order for `on_unsubscribe`

Or even something like the following, assuming the concurrency model allows it:
```ruby
module ApplicationCable
  class Channel < ActionCable::Channel::Base
    around_subscription :activate_tenant
    
    protected
    def activate_tenant(&block)
      current_tenant.activate!(&block)
    end
  end
end
```

I've never played with it directly, but would relying on `ActiveSupport::Callbacks` get us a lot of this for free?  Is that too far for the limited set ActionCable needs?  I'd like to take a stab at this after feedback around this functionality.

**Comment by dhh** [comment]
Hey Mike,

Great to hear you porting an app over. I'd be happy to see us expand the
callbacks. Looking into what ActiveSupport gives us would be great. Sounds
like you want before_subscribe.

**Comment by cristianbica** [comment]
@mieko I've implemented AS::Callbacks for AC::Connection [here](https://github.com/rails/actioncable/compare/master...cristianbica:csrf). Probably a similar approach can be used for channels also. I would implement `{before,after,around}_{un,}subscribe` which will be executed around the `subscribed/unsubscribed` methods. I can create a separate PR for callbacks for connections/channels if you want.

**Comment by mieko** [comment]
@cristianbica The Channel callbacks have been merged in #69.  An extraction of your Connection callbacks in in the CSRF PR would be awesome (I'm needing the functionality myself).  If you're busy, I can submit a PR next week.

**Comment by mieko** [comment]
(Correction: a mobile/caching issue shows me a "merged" that wasn't there, but I still think they'd be a good pair)

**Comment by cristianbica** [comment]
@mieko yeah go ahead and extract the callbacks from there as the CSRF PR won't be merged soon.

**Comment by jeremy** [comment]
👍
